File: 21-merge-two-sorted-lists/merge-two-sorted-lists.cpp
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        ListNode dummy(0);                  // Dummy node to start the merged list
        ListNode* tail = &dummy;
        
        while (list1 && list2) {
            if (list1->val < list2->val) {
                tail->next = list1;         // Link the smaller node
                list1 = list1->next;        // Move pointer forward
            } else {
                tail->next = list2;
                list2 = list2->next;
            }
            tail = tail->next;              // Advance the tail
        }
        // Attach the remaining nodes, if any
        if (list1) tail->next = list1;
        else tail->next = list2;
        
        return dummy.next;                  // Return head of merged list (skip dummy node)
    }
};

File: 21-merge-two-sorted-lists/merge-two-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        n = ListNode()
        m = n 

        

        while list1 or list2:
            if list1 and list2:
                if list1.val > list2.val :
                    n.next = list2 
                    n = n.next
                    list2 = list2.next 
                else :
                    n.next = list1
                    n = n.next
                    list1 = list1.next 
            elif list1:
                n.next = list1
                break 
            elif list2 :
                n.next = list2
                break 
            
        return m.next 



        